builder/sizes_test.go

```diff
@@ -0,0 +1,85 @@
+package builder
+
+import (
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/tinygo-org/tinygo/compileopts"
+)
+
+var sema = make(chan struct{}, runtime.NumCPU())
+
+type sizeTest struct {
+	target     string
+	path       string
+	codeSize   uint64
+	rodataSize uint64
+	dataSize   uint64
+	bssSize    uint64
+}
+
+// Test whether code and data size is as expected for the given targets.
+// This tests both the logic of loadProgramSize and checks that code size
+// doesn't change unintentionally.
+//
+// If you find that code or data size is reduced, then great! You can reduce the
+// number in this test.
+// If you find that the code or data size is increased, take a look as to why
+// this is. It could be due to an update (LLVM version, Go version, etc) which
+// is fine, but it could also mean that a recent change introduced this size
+// increase. If so, please consider whether this new feature is indeed worth the
+// size increase for all users.
+func TestBinarySize(t *testing.T) {
+	// This is a small number of very diverse targets that we want to test.
+	tests := []sizeTest{
+		// microcontrollers
+		{"hifive1b", "examples/echo", 4943, 229, 0, 2252},
+		{"microbit", "examples/serial", 3012, 164, 8, 2248},
+		{"wioterminal", "examples/pininterrupt", 7164, 744, 116, 6424},
+
+		// wasm
+		{"wasm", "examples/serial", 32303, 0, 792, 130280},
+		{"wasi", "examples/serial", 14874, 0, 316, 130756},
+	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.target+"/"+tc.path, func(t *testing.T) {
+			t.Parallel()
+
+			// Build the binary.
+			options := compileopts.Options{
+				Target:        tc.target,
+				Opt:           "z",
+				Semaphore:     sema,
+				InterpTimeout: 60 * time.Second,
+				Debug:         true,
+				VerifyIR:      true,
+			}
+			target, err := compileopts.LoadTarget(&options)
+			if err != nil {
+				t.Fatal("could not load target:", err)
+			}
+			config := &compileopts.Config{
+				Options: &options,
+				Target:  target,
+			}
+			result, err := Build(tc.path, "", t.TempDir(), config)
+			if err != nil {
+				t.Fatal("could not build:", err)
+			}
+
+			// Check whether the size of the binary matches the expected size.
+			sizes, err := loadProgramSize(result.Executable, nil)
+			if err != nil {
+				t.Fatal("could not read program size:", err)
+			}
+			if sizes.Code != tc.codeSize || sizes.ROData != tc.rodataSize || sizes.Data != tc.dataSize || sizes.BSS != tc.bssSize {
```

Comment:
Do we want some fudge factors here (+/- 0.01% or something?) to adjust for instruction selection variations?  This test just seems really fragile with all the possible variables (even the full path name for the debug info..)

Reply:
> Do we want some fudge factors here (+/- 0.01% or something?) to adjust for instruction selection variations?

For the same Go and LLVM versions, the selected instructions should be identical.
Also, it's hard to define "small". A small change in the input source can result in a very large change in the binary, just because some compiler heuristic went wrong. So it's hard to say what kind of fudge factor would be good.

>  This test just seems really fragile with all the possible variables (even the full path name for the debug info..)

I believe that binary size should be identical when using the same LLVM and Go versions (which are only updated every half year in CI), and the same C library versions (musl, wasi-libc, etc, which are updated even less frequently). Debug information is not part of the calculated binary size so changing file paths won't affect the test results.

That said, I do see a failure in CI and it might be an actual bug:

```
    sizes_test.go:78: Unexpected code size when compiling: -target=microbit examples/serial
    sizes_test.go:79:             code rodata   data    bss
    sizes_test.go:80: expected:   3012    164      8   2248
    sizes_test.go:81: actual:     3020    164      8   2248
```

I suspect it's because of some patches in Debian LLVM that affect the binary, I'll need to investigate what's going on here.

Reply:
> I believe that binary size should be identical when using the same LLVM and Go versions (which are only updated every half year in CI), and the same C library versions (musl, wasi-libc, etc, which are updated even less frequently). 

Ah right, so this test is really only useful in the context of CI.  That makes sense.  I wonder if we want to conditionalize it on some environment variable so that running all the tests on dev machines doesn't trigger spurious failures.

Reply:
I believe that as long as the test runs using a self-built LLVM version, the output should be stable at a given TinyGo version.
I've updated the PR to only run the tests with a self-built LLVM, not with a precompiled LLVM (apt.llvm.org, Homebrew, etc). I _hope_ that's good enough.